pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,364 @@
+			// init first query
+			searchResult, err := baseQuery.Do(ctx) // execute
+			if err != nil {
+				level.Error(util.Logger).Log("msg", fmt.Sprintf("Query in index %s met error!", query.TableName))
+				return nil
+			}
+			var batch readBatch
+			var ttyp IndexEntry
+			for _, item := range searchResult.Each(reflect.TypeOf(ttyp)) {
+				if t, ok := item.(IndexEntry); ok {
+					level.Debug(util.Logger).Log("msg", fmt.Sprintf("Index by hash %s: range %s, value %s\n", t.Hash, t.Range, t.Value))
+					batch.rangeValue = []byte(t.Range)
+					batch.value = []byte(t.Value)
+
+					if !callback(&batch) {
+						return nil
+					}
+				}
+			}
+			if searchResult.Hits.TotalHits > int64(e.cfg.MaxFetchDocs) { // need to run "search_after"
+				var sortValue interface{}
+				for {
+					if len(searchResult.Hits.Hits) == 0 {
+						break
+					}
+					//  Get Sort value from last SearchResult
+					sortValue = searchResult.Hits.Hits[len(searchResult.Hits.Hits)-1].Sort[0]
+
+					// Build the query
+					// build base Query again due to limitation of ElasticSearch Client
+					// https://github.com/olivere/elastic/issues/1260
+					// hopefully we can get an elegant way (maybe like baseQuery.SearchAfterClear()) to do this in the future.
+					baseQuery = e.client.Search().Index(query.TableName).Size(e.cfg.MaxFetchDocs).
+						Query(hashTermQuery).Sort("_id", true)
+					if valueTermQuery != nil {
+						baseQuery = baseQuery.Query(valueTermQuery)
+					}
+					if rangeQuery != nil {
+						baseQuery = baseQuery.Query(rangeQuery)
+					}
+					// Search
+					searchResult, err = baseQuery.SearchAfter(sortValue).Do(ctx) // execute
+					if err == io.EOF {
+						break
+					}
+
+					if err != nil {
+						level.Error(util.Logger).Log("msg", fmt.Sprintf("Query in index %s met error!", query.TableName))
+						return nil
+					}
+					var batch readBatch
+					var ttyp IndexEntry
+					for _, item := range searchResult.Each(reflect.TypeOf(ttyp)) {
+						if t, ok := item.(IndexEntry); ok {
+							level.Debug(util.Logger).Log("msg", fmt.Sprintf("Index by hash %s: range %s, value %s\n", t.Hash, t.Range, t.Value))
+							batch.rangeValue = []byte(t.Range)
+							batch.value = []byte(t.Value)
+
+							if !callback(&batch) {
+								return nil
+							}
+						}
+					}
+				}
+			}
+			return nil
+		})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewESIndexClient creates a new IndexClient that used ElasticSearch.
+func NewESIndexClient(cfg Config) (chunk.IndexClient, error) {
+	client, err := newES(cfg)
+	if err != nil {
+		return nil, err
+	}
+	indexClient := &esClient{
+		cfg,
+		client,
+	}
+	return indexClient, nil
+}
+
+func newES(cfg Config) (*elastic.Client, error) {
+	//fix x509: certificate signed by unknown authority
+	var tr *http.Transport
+	if !strings.Contains(cfg.Address, "https") || cfg.TLSSkipVerify {
+		tr = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```

Comment:
so you mean even for http, we'll have option to not skip tls verify?

Reply:
I think I've already have it configurable. If user use an ES addr without "https" or set "tls_skip_verify" to true, it will use the skip verify condition.

Reply:
Yes you did. I must have missed the if statement above